SalesforceCommerceCloud/pwa-kit
# @W-14003846 [MRT Metadata] Fetch nested pwa-kit package versions in a project

# Description

GUS WI: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001Z9zFPYAZ/view

For projects using template extensibility, the versions of the pwa-kit packages (`@salesforce/pwa-kit-dev`, `@salesforce/pwa-kit-runtime`, and `@salesforce/pwa-kit-react-sdk`) are not listed as direct dependencies in the `package.json`. To solve this we need to look it up in the nested dependencies and retrieve the lowest version if it is listed multiple times.

# Types of Changes


> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- Add `getAllProjectDependencyVersions` function to `pwa-kit-dev/src/utils/script-utils.ts`
- Add `getLowestPackageVersion` function to `pwa-kit-dev/src/utils/script-utils.ts`
- Update the `createBundle` function in `pwa-kit-dev/src/utils/script-utils.ts` to get the nested pwa-kit dependency versions if the bundle is using template extensibility

# How to Test-Drive This PR

- Checkout this branch
```
git pull && git checkout upload-nested-package-versions
```
- Create a demo app that uses template extensibility
```
npx @salesforce/pwa-kit-create-app
cd retail-react-app-demo
```
- Verify pwa-kit packages `@salesforce/pwa-kit-dev`, `@salesforce/pwa-kit-runtime`, and `@salesforce/pwa-kit-react-sdk` are not listed under `dependencies` in the `package.json`
- Upload the bundle to cloud-staging
```
npm run build

# Make sure to update PATH_TO_PWA_KIT and USERNAME
PATH_TO_PWA_KIT/pwa-kit/packages/pwa-kit-dev/bin/pwa-kit-dev.js push \
-m "Test nested_pwa-kit packages using @salesforce/retail-react-app" \
-s 'jinsu' \
--credentialsFile /Users/USERNAME/.mobify--cloud-staging.mobify-staging.com \
--cloud-origin https:/cloud-staging.mobify-staging.com
```
- Verify in PBJT that the pwa-kit package versions are listed under bundle_metadata: https://cloud-staging.mobify-staging.com/pbjt/core/bundle/

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
